### PR TITLE
[DI] Add "by-id" autowiring: a side-effect free variant of it based on the class<>id convention

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This PR adds a new autowiring mode, based only on the class <> id convention.
This way of autowiring is free from any conflicting behavior, which is what I was looking for to begin with.

The expected DX is a bit more involving than the current way we do autowiring. But it's worth it to me, because it's plain predictable - a lot less "magic" imho.

So in this mode, for each `App\Foo` type hint, a reference to an "App\Foo" service will be created. If no such service exists, an exception will be thrown. To me, this opens a nice DX: when type hinting interfaces (which is the best practice), this will tell you when you need to create the explicit interface <> id mapping that is missing - thus encourage things to be made explicit, but only when required, and gradually, in a way that will favor discoverability by devs.

Of course, this is opt-in, and BC. You'd need to do eg in yaml: `autowire: by_id`.
For consistency, the current mode (`autowire: true`) can be configured using `autowire: by_type`.

### CODE REVIEW

**Comment by theofidry** [comment]
cool, will try it during the week/weekend!

**Comment by javiereguiluz** [comment]
Could you please show a before/after comparison of what the user has to do to use this feature? Or some small but complete code example? Otherwise it's hard to follow :(  Thanks!

File: src/Symfony/Component/DependencyInjection/Definition.php [hunk 1]
```diff
@@ -737,19 +740,34 @@ public function setAutowiringTypes(array $types)
...
+    public function getAutowired()
+    {
         return $this->autowired;
     }
 
     /**
      * Sets autowired.
      *
-     * @param bool $autowired
+     * @param int $autowired
```
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Definition.php]: `int|false`?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Definition.php]: `int` only, even if bools are accepted, that's only "by cast", which is fine even on php 7.

Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Definition.php]: Imo we should either explicitly allow `false` or add a new constant, it feels weird to call `$definition->setAutowired(0);`.

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Definition.php]: on php 7, adding just `int` would be fine, no BC break, and even then, `false` *would be* accepted.

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Definition.php]: turned to bool|int

**Comment by nicolas-grekas** [comment]
I played with this PR on symfony-demo, I really like the taste it provides :)
I found some enhancements to do made to exception messages: these provide suggestions when the corresponding type is not found as an id.

A typical example is this one, displayed while building an incompletely defined container:
```
  [Symfony\Component\DependencyInjection\Exception\RuntimeException]           
  Cannot autowire argument $manager of method AppBundle\Form\Type\TagsInputTy  
  pe::__construct() for service "AppBundle\Form\Type\TagsInputType": Service   
  "Doctrine\Common\Persistence\ObjectManager" does not exist and it cannot be  
   auto-registered. This type could be aliased to the existing "doctrine.orm.  
  default_entity_manager" service.                                             
```

The resolution is easy, just add the following to services.yml:
```yaml
    Doctrine\Common\Persistence\ObjectManager:
        '@doctrine.orm.default_entity_manager'
```

Of course, this alias should be provided by default by DoctrineBundle. By doing so, we would fix "by-id" autowiring immediatly, but we would also prevent any ambiguous "by-type" autowiring - by stating that "doctrine.orm.default_entity_manager" is the service to use by default when the "Doctrine\Common\Persistence\ObjectManager" type is wired.

A possible working resulting services.yml file for symfony-demo could be the following:
```yaml
services:
    _defaults:
        public: false
        autowire: by_id

    _instanceof:
        Symfony\Component\EventDispatcher\EventSubscriberInterface:
            tags: [ kernel.event_subscriber ]

        Symfony\Component\Form\AbstractType:
            tags: [ form.type ]

        Twig_Extension:
            tags: [ twig.extension ]

    AppBundle\:
        resource: ../../src/AppBundle/{EventListener,Form/Type}

    AppBundle\Controller\:
        resource: ../../src/AppBundle/Controller
        public: true

    AppBundle\Twig\AppExtension:
        arguments: { $locales: '%app_locales%' }

    AppBundle\EventListener\CommentNotificationListener:
        arguments: { $sender: '%app.notifications.email_sender%' }

    AppBundle\EventListener\RedirectToPreferredLocaleListener:
        arguments:
            $locales: '%app_locales%'
            $defaultLocale: '%locale%'

    AppBundle\Security\PostVoter:
        tags: [ security.voter ]

    Twig_Extensions_Extension_Intl:
        class: Twig_Extensions_Extension_Intl

# The following aliases are required for now when using by-id autowiring,
# until these types are aliased in each core bundle, which they should be.
# This kind of explicit aliasing is the only difference between by-id and by-type autowiring
# in this file.

    Doctrine\ORM\EntityManager:
        '@doctrine.orm.default_entity_manager'

    Doctrine\Common\Persistence\ObjectManager:
        '@doctrine.orm.default_entity_manager'

    Swift_Mailer:
        '@swiftmailer.mailer.default'

    Symfony\Component\Routing\Generator\UrlGeneratorInterface:
        '@router.default'
```

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd [hunk 2]
```diff
@@ -279,4 +279,10 @@
       <xsd:pattern value="(%.+%|true|false)" />
     </xsd:restriction>
   </xsd:simpleType>
+
+  <xsd:simpleType name="autowire">
+    <xsd:restriction base="xsd:string">
+      <xsd:pattern value="(true|false|by-type|by-id)" />
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: this does not allow parameters anymore, unlike ``boolean`` (see just above), so it is a BC break.

**Comment by stof** [comment]
for the url generator, please add the alias as this has to be done in FrameworkBundle

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: actually I wanted to trigger this discussion :)
so: in Definition::setAutowired, there is a cast to bool currently.
This means that parameters never worked, did they?
Adding `%..%` to the regexp might fix the BC by still allowing one to pass a param,
but it won't fix the fact that this param is ignored.
Or did I miss something?
Btw, this is the case for all definition setters that "cast" - and has never been an issue really (autowiring/abstract/etc. should not be parametric.)

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 3]
```diff
@@ -257,9 +258,9 @@ private function autowireMethod(\ReflectionMethod $reflectionMethod, array $argu
                 continue;
             }
 
-            $typeName = InheritanceProxyHelper::getTypeHint($reflectionMethod, $parameter, true);
+            $type = InheritanceProxyHelper::getTypeHint($reflectionMethod, $parameter, true);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: As all these renamings are already merged as part of another PR now, can you rebase your branch to make the diff more readable ?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: rebased

**Comment by nicolas-grekas** [comment]
First commit now split as separate PR, adding logging to AutowirePass, see #22095.

This made me realize that by-id autoregistration cannot work, as shown by these kind of log lines, which are obviously wrong:
> Type "Symfony\Component\HttpFoundation\Session\Session" has been auto-registered for service "App\Controller\Actions".

PR updated accordingly. Still very useful to me to help devs wire things explicitly but progressively.

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd (hunk 2 above)
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: if it never worked, this is fine for me.

**Comment by nicolas-grekas** [comment]
Exception message enhanced a bit more, with suggestions for already aliased types:
```
    [Symfony\Component\DependencyInjection\Exception\RuntimeException]             
    Cannot autowire argument $entityManager of method AppBundle\EventListener\C    
    heckRequirementsSubscriber::__construct() for service "AppBundle\EventListe    
    ner\CheckRequirementsSubscriber": Service "Doctrine\ORM\EntityManager" does    
     not exist. This type could be aliased to the existing "doctrine.orm.defaul    
    t_entity_manager" service. This service is already aliased to "Doctrine\ORM    
    \EntityManagerInterface".                                                      
```

**Comment by weaverryan** [comment]
Hi guys!

Let me summarize the new mode! This mode no longer looks through the system and tries to find services that match the type-hint. Instead, it's much simpler: it only looks to see if a service id exists that matches the type-hint. Otherwise, it throws an exception: it does *not* try to automatically create an autowired private service for that class (if it's a class).

I like this! (but I don't like having 2 modes). It doesn't work how I expect autowiring to work (i.e. finding services matching a type-hint), but it's perfectly straightforward. Also, if you take the other mode to an extreme, you would have *zero* services explicitly wired (other than controllers) because the type-hints in classes would cause autowiring to auto-create services. But with this mode, we're saying: all services must be explicitly wired (while in reality for DX, the PSR-4 loader helps a lot here). That *feels* a bit better: I will be able to see in my `services.yml` file why every service exists.

Anyways, I like this... but I don't like the 2 modes. I don't mean we should remove the old one. Instead, if we accept this, it should be with the intention that this becomes *the* recommended mode for autowiring.

Cheers!

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 4]
```diff
@@ -493,10 +497,32 @@ private function createAutowiredDefinition(\ReflectionClass $typeHint)
...
-        $message = sprintf('No services were found matching the "%s" %s and it cannot be auto-registered', $type, $classOrInterface);
+        if (null === $this->types) {
+            $this->populateAvailableTypes();
+        }
+        if (!$autowireById) {
+            $message = sprintf('No services were found matching the "%s" %s and it cannot be auto-registered', $type, $classOrInterface);
+        } elseif (isset($this->ambiguousServiceTypes[$type])) {
+            $message = sprintf('Service "%s" does not exist. This type could be aliased to any of these existing services: "%s"', $type, implode('", "', $this->ambiguousServiceTypes[$type]));
+        } elseif (isset($this->types[$type])) {
+            $message = sprintf('Service "%s" does not exist. This type could be aliased to the existing "%s" service', $type, $this->types[$type]);
```
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: It's perhaps outside the scope of this PR, but we could also add this "alias" recommendation to the "other" mode, when there are multiple definitions (https://github.com/symfony/symfony/blob/daac6c24490cfd9eb995ce6fb69095b7241f8b13/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php#L56)

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 5]
```diff
@@ -493,10 +497,32 @@ private function createAutowiredDefinition(\ReflectionClass $typeHint)
...
+
+            $aliases = array();
+            foreach (array_keys($this->types, $this->types[$type]) as $type) {
+                if ($this->container->has($type)) {
+                    $aliases[] = $type;
+                }
+            }
+            if ($aliases) {
+                $message .= sprintf('. This service is already aliased to "%s"', implode('", "', $aliases));
+            }
```
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: What case is this catching? I don't see anything in `AutowirePassTest` for this.

**Comment by theofidry** [comment]
Wholeheartedly agree with @weaverryan 😄

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: See my previous comment, last sentence of the block:
https://github.com/symfony/symfony/pull/22060#issuecomment-288232673
This PR has no tests yet :)

Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Ah, thank you! What about:

> . Or you can update your type hint to "%s".

(but, if we did this, we probably need some extra logic to make the sentence read correctly if there are multiple `$aliases`):

> . Or you can update your type hint to be one of the following: "%s".

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: @weaverryan see #22135

**Comment by nicolas-grekas** [comment]
Test case added, failure unrelated. Ready.

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by jordscream** [comment]
@nicolas-grekas , you think you update the doc soon ^^ ? need it to prevent my partner using this implementation when autowire fails causing by service definition priority.
thks for all and good work 👍

**Comment by curry684** [comment]
http://symfony.com/blog/new-in-symfony-3-3-configurable-autowiring-setters is still live without changes, which I was trying last Friday before this PR got merged, and cost me a lot of time as `autowire: [ setX ]` was still accepted at the time but not doing anyway. Current dev-master throws an error on that thankfully, but the blog announcement is still up.

**Comment by dunglas** [comment]
Is this new mode still necessary now that we have #22254? IIUC, they fix almost the same issue (multiple implementations of an interface). For classes without interfaces, the default mode will less likely fail and is faster/easier to use from a dev POV.

Anyway, this new mode can also be useful in some cases (e.g. the domain layer - it's an intermediate between 100% manual wiring and the "real" autowiring mode) but IMO it should not be the default.
